src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -1285,21 +1285,149 @@ internal bool HasCollectionExpressionApplicableAddMethod(SyntaxNode syntax, Type
             }
         }
 
+        internal bool HasCollectionExpressionAddMethod(SyntaxNode syntax, TypeSymbol targetType)
+        {
+            const string methodName = "Add";
+            var useSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded;
+
+            var implicitReceiver = new BoundObjectOrCollectionValuePlaceholder(syntax, isNewInstance: true, targetType) { WasCompilerGenerated = true };
+            var memberAccess = BindInstanceMemberAccess(
+                syntax,
+                right: syntax,
+                boundLeft: implicitReceiver,
+                rightName: methodName,
+                rightArity: 0,
+                typeArgumentsSyntax: default,
+                typeArgumentsWithAnnotations: default,
+                invoked: true,
+                indexed: false,
+                BindingDiagnosticBag.Discarded);
+            if (memberAccess is not BoundMethodGroup methodGroup)
+            {
+                return false;
+            }
+
+            Debug.Assert(methodGroup.SearchExtensionMethods);
+
+            if (methodGroup.ResultKind == LookupResultKind.Viable &&
+                methodGroup.Methods.Any(m => isApplicableAddMethod(m.GetConstructedLeastOverriddenMethod(accessingTypeOpt: null, requireSameReturnType: false), expectingExtensionMethod: false)))
+            {
+                return true;
+            }
+
+            if (methodGroup.SearchExtensionMethods &&
+                hasApplicableAddExtensionMethod(syntax, implicitReceiver))
+            {
+                return true;
+            }
+
+            return false;
+
+            bool hasApplicableAddExtensionMethod(SyntaxNode syntax, BoundObjectOrCollectionValuePlaceholder implicitReceiver)
```

Comment:
>hasApplicableAddExtensionMethod [](http://example.com/codeflow?start=17&length=31)

Consider adding a comment that ```Binder.BindExtensionMethod``` is a "prototype" for the logic in this local function